### PR TITLE
[Client][Manager] Support http2 connection, show IP and http version on manager

### PR DESCRIPTION
**Pull request by talregev**
Add support http2 connection. Client can connect to the server and account manager in http2.
Show IP and http version on manager.

Currently World Community Grid and Boinc Account Manager are support http2. 

Test it on windows 11 and wsl2 ubuntu 22.04.

image

image

image

wsl2 ubuntu 22.04:
image

### CODE REVIEW

**Review by AenBleidd — CHANGES_REQUESTED** [review]
While adding support of http2 to Windows build is generally fine, looks like no other OS is touched by this PR.

@talregev, does that mean that Linux, OSX and Android already have support of it and only in Windows build it was missed?

But in any case, this information is useless for the most of users, because it is usually static and don't represent anything that user can fix/improve.

**Comment by talregev** [comment]
In this PR It effect Windows and Linux.
I can add to here for android without visible notice for the user. 
For Mac I will add support in following PR. 

It another info about server and transformation. Like the name of the project is static and user cannot effect it and many other information we display for the user.

**Comment by AenBleidd** [comment]
> It another info about server and transformation. Like the name of the project is static and user cannot effect it and many other information we display for the user.

'Project Name' is static, that's true. But it shows to the user information what is the Project application tries to download/upload from/to. And in case smth stuck - user has possibility to contact the Project.
But neither http version not ip version don't give any valuable information that user could use in any way: project either support http2 or not, and if not - user can do nothing to change that. The same is about the ip version.
That is why this is useless information.

As you may remember, we try to keep UI as simple as possible and provide only valuable information to our users (and still some of them think it's to techy and it should be made simpler and less boring).
Adding http version and ip version (that most of the users have no idea what does that mean and what's the difference between them) overloads UI with information that is not needed.

I see no case when this information could be useful.

I'm sorry, but as a general advice: before starting to implement smth that has no open ticket - ask first, because there could be a reason (like this) why this feature is not welcome.

**Comment by talregev** [comment]
User also can contact to Project and ask about planning to add support in http2 that can lead to stable transformations. 
Also without noticeable UI, you cannot test this feature easily. For example for android.

**Comment by talregev** [comment]
Currently is not working on vcpkg Linux. I am not sure why. Can you test it on your side?

**Comment by AenBleidd** [comment]
> Currently is not working on vcpkg Linux. I am not sure why. Can you test it on your side?

I don't think I have time for this right now.
Meanwhile it would be good if you:
- Verify that you don't have `http_1_0` set to `1` in your `cc_config.xml` file
- Set `http_debug` logging flag and provide logs

Also, you can temporary make `libcurl_exec()` function public and test it separately with debug logs enabled (from the previous point) and test download on some page with http2 support (e.g. http://www.http2demo.io/)

**Comment by talregev** [comment]
> Currently is not working on vcpkg Linux. I am not sure why. Can you test it on your side?

I fix the bug.

File: client/pers_file_xfer.cpp [hunk 1]
```diff
@@ -451,12 +451,16 @@ int PERS_FILE_XFER::write(MIOFILE& fout) {
             "        <file_offset>%f</file_offset>\n"
             "        <xfer_speed>%f</xfer_speed>\n"
             "        <url>%s</url>\n"
+            "        <http_version>%s</http_version>\n"
```
Comment (AenBleidd) [inline client/pers_file_xfer.cpp:454]: These two lines are not needed, please remove

File: clientgui/ViewTransfers.cpp [hunk 2]
```diff
@@ -50,6 +50,8 @@
 #define COLUMN_TOCOMPLETION         5
 #define COLUMN_SPEED                6
 #define COLUMN_STATUS               7
+#define COLUMN_HTTP_VERSION         8
```
Comment (AenBleidd) [inline clientgui/ViewTransfers.cpp:53]: These two lines are not needed, please remove

File: clientgui/ViewTransfers.cpp [hunk 3]
```diff
@@ -59,7 +61,7 @@
 // For now, show all columns by default
 static int DefaultShownColumns[] = { COLUMN_PROJECT, COLUMN_FILE, COLUMN_PROGRESS, 
                                 COLUMN_SIZE, COLUMN_TIME, COLUMN_TOCOMPLETION, COLUMN_SPEED,
-                                COLUMN_STATUS};
+                                COLUMN_STATUS, COLUMN_HTTP_VERSION, COLUMN_IP_VERSION};
```
Comment (AenBleidd) [inline clientgui/ViewTransfers.cpp:64]: This is not needed, please remove

File: clientgui/ViewTransfers.cpp [hunk 4]
```diff
@@ -86,6 +88,8 @@ CTransfer::~CTransfer() {
     m_strTime.Clear();
     m_strTimeToCompletion.Clear();
     m_strSpeed.Clear();
+    m_strHttpVersion.Clear();
```
Comment (AenBleidd) [inline clientgui/ViewTransfers.cpp:91]: These two lines are not needed, please remove

File: clientgui/ViewTransfers.cpp [hunk 5]
```diff
@@ -172,6 +176,11 @@ static bool CompareViewTransferItems(int iRowIndex1, int iRowIndex2) {
     case COLUMN_STATUS:
 	result = transfer1->m_strStatus.CmpNoCase(transfer2->m_strStatus);
         break;
+    case COLUMN_HTTP_VERSION:
```
Comment (AenBleidd) [inline clientgui/ViewTransfers.cpp:179]: These are not needed, please remove

File: clientgui/ViewTransfers.cpp [hunk 6]
```diff
@@ -230,6 +239,8 @@ CViewTransfers::CViewTransfers(wxNotebook* pNotebook) :
     m_aStdColNameOrder->Insert(_("Remaining (estimated)"), COLUMN_TOCOMPLETION);
     m_aStdColNameOrder->Insert(_("Speed"), COLUMN_SPEED);
     m_aStdColNameOrder->Insert(_("Status"), COLUMN_STATUS);
+    m_aStdColNameOrder->Insert(_("HTTP Version"), COLUMN_HTTP_VERSION);
```
Comment (AenBleidd) [inline clientgui/ViewTransfers.cpp:242]: These two lines are not needed, please remove

File: clientgui/ViewTransfers.cpp [hunk 7]
```diff
@@ -246,6 +257,8 @@ CViewTransfers::CViewTransfers(wxNotebook* pNotebook) :
     m_iStdColWidthOrder.Insert(100, COLUMN_TOCOMPLETION);
     m_iStdColWidthOrder.Insert(80, COLUMN_SPEED);
     m_iStdColWidthOrder.Insert(150, COLUMN_STATUS);
+    m_iStdColWidthOrder.Insert(100, COLUMN_HTTP_VERSION);
```
Comment (AenBleidd) [inline clientgui/ViewTransfers.cpp:260]: These two lines are not needed, please remove

File: clientgui/ViewTransfers.cpp [hunk 8]
```diff
@@ -303,6 +316,14 @@ void CViewTransfers::AppendColumn(int columnID){
             m_pListPane->AppendColumn((*m_aStdColNameOrder)[COLUMN_STATUS],
                 wxLIST_FORMAT_LEFT, m_iStdColWidthOrder[COLUMN_STATUS]);
             break;
+        case COLUMN_HTTP_VERSION:
```
Comment (AenBleidd) [inline clientgui/ViewTransfers.cpp:319]: These are not needed. please remove

File: clientgui/ViewTransfers.cpp [hunk 9]
```diff
@@ -529,6 +550,12 @@ wxString CViewTransfers::OnListGetItemText(long item, long column) const {
         case COLUMN_STATUS:
             strBuffer = transfer->m_strStatus;
             break;
+        case COLUMN_HTTP_VERSION:
```
Comment (AenBleidd) [inline clientgui/ViewTransfers.cpp:553]: These are not needed, please remove

File: clientgui/ViewTransfers.cpp [hunk 10]
```diff
@@ -680,6 +707,20 @@ bool CViewTransfers::SynchronizeCacheItem(wxInt32 iRowIndex, wxInt32 iColumnInde
                 return true;
             }
             break;
+        case COLUMN_HTTP_VERSION:
```
Comment (AenBleidd) [inline clientgui/ViewTransfers.cpp:710]: These are not needed, please remove

File: clientgui/ViewTransfers.cpp [hunk 11]
```diff
@@ -718,6 +759,38 @@ void CViewTransfers::GetDocFileName(wxInt32 item, wxString& strBuffer) const {
 }
 
 
+void CViewTransfers::GetDocHTTPVersion(wxInt32 item, wxString& strBuffer) const {
```
Comment (AenBleidd) [inline clientgui/ViewTransfers.cpp:762]: This function is not needed, please remove

File: clientgui/ViewTransfers.cpp [hunk 12]
```diff
@@ -718,6 +759,38 @@ void CViewTransfers::GetDocFileName(wxInt32 item, wxString& strBuffer) const {
...
+
+    if (transfer) {
+        strBuffer = wxString(transfer->http_version.c_str(), wxConvUTF8);
+    } else {
+        strBuffer = wxEmptyString;
+    }
+}
+
+
+void CViewTransfers::GetDocIPVersion(wxInt32 item, wxString& strBuffer) const {
```
Comment (AenBleidd) [inline clientgui/ViewTransfers.cpp:778]: This function is not needed, please remove

File: clientgui/ViewTransfers.h [hunk 13]
```diff
@@ -46,6 +46,8 @@ class CTransfer : public wxObject {
     wxString m_strTime;
     wxString m_strTimeToCompletion;
     wxString m_strSpeed;
+    wxString m_strHttpVersion;
```
Comment (AenBleidd) [inline clientgui/ViewTransfers.h:49]: These two lines are not needed, please remove

File: clientgui/ViewTransfers.h [hunk 14]
```diff
@@ -102,6 +104,8 @@ class CViewTransfers : public CBOINCBaseView {
     wxInt32                 FormatSpeed( double fBuffer, wxString& strBuffer ) const;
     void                    GetDocStatus(wxInt32 item, wxString& strBuffer) const;
     void                    GetDocProjectURL(wxInt32 item, wxString& strBuffer) const;
+    void                    GetDocHTTPVersion(wxInt32 item, wxString& strBuffer) const;
```
Comment (AenBleidd) [inline clientgui/ViewTransfers.h:107]: These two functions are not needed, please remove

File: lib/gui_rpc_client.h [hunk 15]
```diff
@@ -336,6 +336,8 @@ struct FILE_TRANSFER {
     std::string hostname;
     double project_backoff;
     PROJECT* project;
+    std::string http_version;
```
Comment (AenBleidd) [inline lib/gui_rpc_client.h:339]: These two lines are not needed, please remove

File: lib/gui_rpc_client_ops.cpp [hunk 16]
```diff
@@ -790,6 +790,8 @@ int FILE_TRANSFER::parse(XML_PARSER& xp) {
         if (xp.parse_double("xfer_speed", xfer_speed)) continue;
         if (xp.parse_string("hostname", hostname)) continue;
         if (xp.parse_double("project_backoff", project_backoff)) continue;
+        if (xp.parse_string("http_version", http_version)) continue;
```
Comment (AenBleidd) [inline lib/gui_rpc_client_ops.cpp:793]: These two lines are not needed, please remove

File: lib/gui_rpc_client_ops.cpp [hunk 17]
```diff
@@ -817,6 +819,8 @@ void FILE_TRANSFER::clear() {
     hostname.clear();
     project = NULL;
     project_backoff = 0;
+    http_version.clear();
```
Comment (AenBleidd) [inline lib/gui_rpc_client_ops.cpp:822]: These two lines are not needed, please remove

File: lib/gui_rpc_client_print.cpp [hunk 18]
```diff
@@ -210,6 +210,8 @@ void FILE_TRANSFER::print() {
     if (xfer_active) printf("   estimated_xfer_time_remaining: %f\n", estimated_xfer_time_remaining);
     printf("   bytes_xferred: %f\n", bytes_xferred);
     printf("   xfer_speed: %f\n", xfer_speed);
+    printf("   http_version: %s\n", http_version.c_str());
```
Comment (AenBleidd) [inline lib/gui_rpc_client_print.cpp:213]: These two lines are not needed, please remove

File: linux/ci_configure_manager.sh [hunk 19]
```diff
@@ -15,4 +15,4 @@ linux/update_vcpkg_manager.sh
 
 export _libcurl_pc="$VCPKG_DIR/lib/pkgconfig/libcurl.pc"
 export PKG_CONFIG_PATH=$VCPKG_DIR/lib/pkgconfig/
-./configure --enable-vcpkg --disable-server --disable-client --with-wx-config=$VCPKG_DIR/tools/wxwidgets/wx-config CPPFLAGS=-DwxDEBUG_LEVEL=0 GTK_LIBS="`pkg-config --libs gtk+-3.0`"
+./configure --enable-vcpkg --disable-server --disable-client --with-wx-config=$VCPKG_DIR/tools/wxwidgets/wx-config GTK_LIBS="`pkg-config --libs gtk+-3.0`"
```
Comment (AenBleidd) [inline linux/ci_configure_manager.sh:18]: This looks like not related to this particular PR.
Please move it from here to a separate PR

File: client/http_curl.cpp [hunk 20]
```diff
@@ -89,6 +89,7 @@ size_t libcurl_write(void *ptr, size_t size, size_t nmemb, HTTP_OP* phop) {
         );
     }
     phop->bytes_xferred += (double)(stWrite);
+    phop->update_http_ip_version();
```
Comment (AenBleidd) [inline client/http_curl.cpp:92]: This function will write to the log file om every write call. This is not needed.
This should be written only once when connections is established

File: client/http_curl.cpp [hunk 21]
```diff
@@ -148,6 +149,7 @@ size_t libcurl_read(void *ptr, size_t size, size_t nmemb, HTTP_OP* phop) {
         phop->bytes_xferred += (double)(stRead);
         daily_xfer_history.add(stRead, true);
     }
+    phop->update_http_ip_version();
```
Comment (AenBleidd) [inline client/http_curl.cpp:152]: This function will write to the log file om every read call. This is not needed.
This should be written only once when connections is established

File: client/http_curl.h [hunk 22]
```diff
@@ -60,6 +60,8 @@ class HTTP_OP {
     char m_url[1024];  
     char m_curl_user_credentials[1024];
         // string needed for proxy username/password
+    char m_http_version[100];
```
Comment (AenBleidd) [inline client/http_curl.h:63]: This information not need to be stored, so please remove these two member variables

File: client/http_curl.cpp [hunk 23]
```diff
@@ -257,6 +292,8 @@ HTTP_OP::HTTP_OP() {
     lSeek = 0;
     xfer_speed = 0;
     is_background = false;
+    safe_strcpy(m_http_version, "");
```
Comment (AenBleidd) [inline client/http_curl.cpp:295]: No need to save this information, so please remove these two lines

File: client/http_curl.cpp [hunk 24]
```diff
@@ -210,6 +212,40 @@ int libcurl_debugfunction(
     return 0;
 }
 
+std::string get_http_version(int _enum) {
```
Comment (AenBleidd) [inline client/http_curl.cpp:215]: According to the documentation ( https://curl.se/libcurl/c/CURLINFO_HTTP_VERSION.html ) valid values are:
- CURL_HTTP_VERSION_1_0
- CURL_HTTP_VERSION_1_1
- CURL_HTTP_VERSION_2_0
- CURL_HTTP_VERSION_3
- or 0 if the version cannot be determined.

See no reason to check for other values

File: client/http_curl.cpp [hunk 25]
```diff
@@ -210,6 +212,40 @@ int libcurl_debugfunction(
...
+    case CURL_HTTP_VERSION_2TLS:
+        return "http/2tls";
+    case CURL_HTTP_VERSION_2_PRIOR_KNOWLEDGE:
+        return "http/2prior_knowledge";
+    case CURL_HTTP_VERSION_3:
+        return "http/3";
+    case CURL_HTTP_VERSION_LAST:
+        return "CURL_HTTP_VERSION_LAST";
+    default:
+        return "HTTP_ERROR";
```
Comment (AenBleidd) [inline client/http_curl.cpp:234]: Better to return `unknown http version`

**Review by AenBleidd — CHANGES_REQUESTED** [review]

**Comment by AenBleidd** [comment]
As it was stated above, neither HTTP2 nor HTTP3 will bring a lot of benefit to the BOINC.
BOINC Client does not send a lot of requests to the servers, so latency is not an issue in our case.
Also, we're usually not sending several parallel requests to the server, and we're even doing max two concurrent file download/upload.

I'm closing this PR without merge.

**Pull request closed without merging**